components/collections/codepointtrie_builder/src/native.rs

```diff
@@ -0,0 +1,152 @@
+where
+    T: TrieValue + Into<u32>,
+{
+    let mut error = 0;
+    let builder = unsafe {
+        umutablecptrie_open_72(
+            cpt_builder.default_value.into(),
+            cpt_builder.error_value.into(),
+            &mut error,
+        )
+    };
+
+    if error != 0 {
+        panic!("cpt builder returned error code {}", error);
+    }
+
+    let CodePointTrieBuilderData::ValuesByCodePoint(values) = cpt_builder.data;
+
+    for (cp, value) in values.iter().enumerate() {
+        unsafe {
+            umutablecptrie_set_72(builder, cp as u32, (*value).into(), &mut error);
+        }
+        if error != 0 {
+            panic!("cpt builder returned error code {}", error);
+        }
+    }
+
+    let trie_type = match cpt_builder.trie_type {
+        TrieType::Fast => 0,
+        TrieType::Small => 1,
+    };
+    let width = match mem::size_of::<T::ULE>() {
+        2 => 0, // UCPTRIE_VALUE_BITS_16
+        4 => 1, // UCPTRIE_VALUE_BITS_32
+        1 => 2, // UCPTRIE_VALUE_BITS_8
+        other => panic!("Don't know how to make trie with width {other}"),
+    };
+    let built = unsafe { umutablecptrie_buildImmutable_72(builder, trie_type, width, &mut error) };
+    if error != 0 {
+        panic!("cpt builder returned error code {}", error);
+    }
+    unsafe {
+        umutablecptrie_close_72(builder);
+    }
+
+    let trie = unsafe { &*built };
+
+    let header = CodePointTrieHeader {
+        high_start: trie.highStart,
+        shifted12_high_start: trie.shifted12HighStart,
+        index3_null_offset: trie.index3NullOffset,
+        data_null_offset: trie
+            .dataNullOffset
+            .try_into()
+            .expect("Found negative or out of range dataNullOffset"),
+        null_value: trie.nullValue,
+        trie_type: TrieType::try_from(trie.type_ as u8).expect("Found out of range TrieType"),
+    };
+
+    let index_slice = unsafe { slice::from_raw_parts(trie.index, trie.indexLength as usize) };
+    let index_vec = ZeroVec::alloc_from_slice(index_slice);
+    let data_vec: Result<Vec<T>, _> = unsafe {
+        match mem::size_of::<T::ULE>() {
+            1 => slice::from_raw_parts(trie.data.ptr8, trie.dataLength as usize)
+                .iter()
+                .map(|x| TrieValue::try_from_u32(*x as u32))
+                .collect(),
+            2 => slice::from_raw_parts(trie.data.ptr16, trie.dataLength as usize)
+                .iter()
+                .map(|x| TrieValue::try_from_u32(*x as u32))
+                .collect(),
+            4 => slice::from_raw_parts(trie.data.ptr32, trie.dataLength as usize)
+                .iter()
+                .map(|x| TrieValue::try_from_u32(*x as u32))
+                .collect(),
+            other => panic!("Don't know how to make trie with width {other}"),
+        }
+    };
+    let data_vec = ZeroVec::alloc_from_slice(&data_vec.map_err(|_| ()).expect("Failed to parse as TrieValue"));
```

Comment:
Question: why are you cleaning out the error? Is the `try_from_u32` error not printable or something?

Reply:
yep, there's no Debug bound